js/hang/src/publish/source/file.ts

```diff
@@ -0,0 +1,128 @@
+import { Signal, Effect } from "@kixelated/signals";
+import type { StreamTrack as VideoStreamTrack } from "../video/types";
+import type { StreamTrack as AudioStreamTrack } from "../audio/types";
+
+export interface FileSourceConfig {
+	enabled: Signal<boolean>;
+}
+
+export class File {
+	#file = new Signal<globalThis.File | undefined>(undefined);
+	signals = new Effect();
+
+	source = new Signal<{ video?: VideoStreamTrack; audio?: AudioStreamTrack }>(
+		{}
+	);
+	enabled: Signal<boolean>;
+
+	constructor(config: FileSourceConfig) {
+		this.enabled = config.enabled;
+
+		this.signals.effect((effect) => {
+			const file = effect.get(this.#file);
+			const enabled = effect.get(this.enabled);
+
+			if (!file || !enabled) {
+				effect.set(this.source, {}, {});
```

Comment:
```suggestion
				this.source.set({});
```

`effect.set` sets the value to the first arg and on cleanup, sets the value to the second arg. We don't need the latter.